### ISSUE TITLE
Split premake & trigger coverage into separate config options (WAS: drop_partition_id() doesn't recreate trigger functions which mention the dropped partition)

### ISSUE THREAD

**Issue by IMSoP** [keithf4/pg_partman#62]
When partitions are dropped by run_maintenance() / drop_partition_id(), the corresponding trigger function is not rebuilt, so still contains references to the dead partition. 

This means that new inserts within the non-retained range have rather unwelcome behaviour: If the table is just disinherited, the insert will proceed, but not be visible afterwards. If the table is dropped or moved, the trigger function will raise an error.

Example: the following will produce `ERROR: relation "public.test_p0" does not exist`:

```
create table test ( id serial, value text not null );

select partman.create_parent('public.test', 'id', 'id', '100');

update partman.part_config set partition_interval = '100'::text;
update partman.part_config set retention = '200'::text;
update partman.part_config set retention_keep_table = false;

insert into test ( value )
select 'test1: ' || generate_series(0, 500);

select partman.run_maintenance();

select min(id), max(id) from test;

insert into test ( id, value )
values ( 10, 'invalid' );
```

**Comment by IMSoP** [comment]
Possible fix. This was tested with the above code only, against a fork of 2.0.0 hacked to run on PG 9.1, hence a patch not a PR.

```
From f046bd38f3cb0740d3b164f000be332f15756d55 Mon Sep 17 00:00:00 2001
From: Rowan Collins <git@rwec.co.uk>
Date: Thu, 18 Jun 2015 20:22:29 +0100
Subject: [PATCH] Simple fix for Issue #62; tested with simple 'id'
 partitioning only

---
 sql/functions/drop_partition_id.sql | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/sql/functions/drop_partition_id.sql b/sql/functions/drop_partition_id.sql
index 63bc970..057d4e1 100644
--- a/sql/functions/drop_partition_id.sql
+++ b/sql/functions/drop_partition_id.sql
@@ -190,6 +190,11 @@ LOOP
 
 END LOOP; -- End child table loop
 
+-- If we've dropped one or more partitions, re-generate the trigger function, which may have it hard-coded
+IF v_drop_count > 0 THEN
+    PERFORM @extschema@.create_function_id(p_parent_table);
+END IF;
+
 IF v_jobmon_schema IS NOT NULL THEN
     IF v_job_id IS NOT NULL THEN
         v_step_id := add_step(v_job_id, 'Finished partition drop maintenance');
-- 
1.9.5.msysgit.0
```

**Comment by keithf4** [comment]
I think the better thing to do here would be one of two things:

- Adjust your premake value to be lower. Assuming you've got the default value of 4 here, you've set the retention value lower than the premake would assume there would be child tables.

- Adjust your retention period to not remove tables that are within the premake window. This would probably be the better choice.

The place to recreate the function would go in the run_maintenance() function. I'll see about adding it, but this is a weird edge case where you're dropping old tables much sooner than expected.

**Comment by IMSoP** [comment]
Fair point that this is a slightly "exotic" situation. 

I came upon it because the changes I'm working on involve having all tables listed in the trigger function, not just current +/- premake. (This is so that I can have partitions of different sizes within the same set.)

It does seem a bit weird to have a link between premake, trigger optimisation, and retention limits though. Why should the number of partitions I create in advance (premake) determine which "past" partitions are optimised in the trigger?

As for where the change should go, `drop_partition_id` is documented as a function which can be called directly, although it is "recommended" to call it via `run_maintenance`. You might for instance want to perform a one-off maintenance task with a very low retention value, and be quite likely to run into this issue as a result. Is there a reason *not* to recreate the function there?

**Comment by keithf4** [comment]
Fair point on where the recreation is. I just looked at where the recreation was when initially creating a partition and it's in run_maintenance. Will revisit it there.

I tied them together because it was easy to do at the time vs making multiple config options. I may consider making a separate trigger optimization value now, tho. I hadn't realized the retention limit relation before. I doubt I'd make a separate one for retention, though, since I think it would be a bit more obvious that it's related to the optimization value.

**Comment by keithf4** [comment]
The issue with retention not recreating the trigger function has been fixed in 2.1.0. Sorry for the delay. Other features took longer than expected. 
Renamed the subject of this issue to reflect the requested feature.

**Comment by keithf4** [comment]
Version 2.2.0 is out and now has distinct configuration options for trigger coverage and constraint management. Thanks for pushing the idea!
